### PR TITLE
[FLINK-21731] Add benchmarks for DefaultScheduler's creation, scheduling and deploying

### PR DESCRIPTION
**Pull request by Thesharing**
## What is the purpose of the change

*This pull request introduces benchmarks to evaluate the performance of DefaultScheduler's creation, scheduling and deploying`.*

*For more details please check [FLINK-21731](https://issues.apache.org/jira/browse/FLINK-21731).*

## Brief change log

  - *Introduces benchmarks that evaluates the performance of creating DefaultScheduler.*
  - *Introduces benchmarks that evaluates the performance of scheduling and deploying in DefaultScheduler.*

## Verifying this change

*This change added the unit tests `CreateSchedulerBenchmarkTest` and `SchedulingAndDeployingBenchmarkTest`.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by Thesharing** [comment]
@flinkbot run azure

**Comment by Thesharing** [comment]
@flinkbot run azure

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulingAndDeployingBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * The benchmark of scheduling and deploying tasks in a STREAMING/BATCH job. The related method is
+ * {@link DefaultScheduler#startScheduling}.
+ */
+public class SchedulingAndDeployingBenchmark extends SchedulerBenchmarkBase {
+
+    private DefaultScheduler scheduler;
+
+    public void setup(JobConfiguration jobConfiguration) throws Exception {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulingAndDeployingBenchmark.java:44]: Looks to me that the `mainThreadExecutor`, `jobGraph `, `physicalSlotProvider ` can all be part of the `SchedulerBenchmarkBase`.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                slotPool,
+                new RpcTaskManagerGateway(testingTaskExecutorGateway, JobMasterId.generate()),
+                slotPoolSize,
+                mainThreadExecutor);
+
+        return new PhysicalSlotProviderImpl(
+                LocationPreferenceSlotSelectionStrategy.createDefault(), slotPool);
+    }
+
+    static void offerSlots(
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java:109]: can be private

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                        jobConfiguration, jobVertices.size(), mainThreadExecutor);
+    }
+
+    public void teardown() throws Exception {
+        if (scheduledExecutorService != null) {
+            scheduledExecutorService.shutdownNow();
+        }
+    }
+
+    static PhysicalSlotProvider createPhysicalSlotProvider(
```
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java:89]: can be private

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java (hunk 2 above)
Comment (Thesharing) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerBenchmarkBase.java:109]: Sorry for being careless. Resolved.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing the comments. @Thesharing 
The change looks good to me.
Merging.

**Comment by tillrohrmann** [comment]
What are these tests actually testing @zhuzhurk @Thesharing? It seems that they only call `SchedulerBase.startScheduling` and that's it. This seems not like a very specific test.

**Comment by Thesharing** [comment]
> What are these tests actually testing @zhuzhurk @Thesharing? It seems that they only call `SchedulerBase.startScheduling` and that's it. This seems not like a very specific test.

We add this benchmark because we find that existing benchmarks cannot cover the entire procedure of scheduling, like `DefaultScheduler#allocateSlotsAndDeploy`, and etc. We cannot add benchmarks for each method, since it will make the benchmarks complicated and takes long time to run them all. Furthermore, if anyone insert a procedure, we cannot track on it. Thus, we think it's necessary to add a benchmark that covers the entire procedure. If anything related to scheduling changes, we can immediately know about it according to this end-to-end benchmark.

**Comment by zhuzhurk** [comment]
The case created a `PhysicalSlotProvider` with all required slots assigned early before starting scheduling.
So that one `startScheduling()` can trigger the process of task scheduling, slot allocation (from SlotPool within JM) and task deployment. It can cover the scheduling/deploying performance of a streaming job (except for the time to allocate slots from RM/TM).

**Comment by tillrohrmann** [comment]
If it is a benchmark, where do we measure its performance? Moreover, where can I see whether it is now slower or faster?

**Comment by Thesharing** [comment]
> If it is a benchmark, where do we measure its performance? Moreover, where can I see whether it is now slower or faster?

It's located at [flink-benchmark](https://github.com/apache/flink-benchmarks/tree/master/src/main/java/org/apache/flink/scheduler/benchmark). We can track its performance at the [codespeed](http://codespeed.dak8s.net:8000/timeline/#/?exe=5&env=2&revs=1000&equid=off&quarts=on&extr=on&ben=grid).

**Comment by tillrohrmann** [comment]
Ah ok, this is good to know. Why do we have additionally the tests for the benchmarks?

**Comment by Thesharing** [comment]
> Ah ok, this is good to know. Why do we have additionally the tests for the benchmarks? Is the intention to ensure that the benchmarks are running?

Yes, these tests make sure that when methods used by the benchmarks are modified, the benchmarks will be fixed, too.

**Comment by tillrohrmann** [comment]
> > Ah ok, this is good to know. Why do we have additionally the tests for the benchmarks? Is the intention to ensure that the benchmarks are running?
> 
> Yes, these tests make sure that when methods used by the benchmarks are modified, the benchmarks will be fixed, too.

Alright, this makes sense. Maybe the next time we can add a small comment stating this purpose. Thanks for the clarification.

**Comment by Thesharing** [comment]
> > > Ah ok, this is good to know. Why do we have additionally the tests for the benchmarks? Is the intention to ensure that the benchmarks are running?
> > 
> > 
> > Yes, these tests make sure that when methods used by the benchmarks are modified, the benchmarks will be fixed, too.
> 
> Alright, this makes sense. Maybe the next time we can add a small comment stating this purpose. Thanks for the clarification.

Sorry for making this confusing. I'll note this the next time.

**Pull request closed without merging**